source/magpie/dataframe.d

```diff
@@ -3418,3 +3688,115 @@ unittest
     df[["Hello", "Hi", "Hey"], 0] = df.asSlice!(Universal)(["Hi", "Hello", "Hello"]);
     assert(df.asSlice!(Universal)(["Hello", "Hi", "Hey"]) == ["0", "2", "2", "nan", "2"]);
 }
+
+// Aggregate Operation on DataFrame Columns
+unittest
+{
+    DataFrame!(int, 2, double, 2) df;
+    Index inx;
+
+    inx[0] = ["Row1", "Row2"];
+    inx[1] = ["Col1", "Col2", "Col3", "Col4"];
+
+    df.setFrameIndex(inx);
+    df = [[1, 2, 3.4, 5.6], [2, 8, 7.9, 5.6]];
+    // df.display();
+
+    import std.algorithm: max, min;
+    assert(df.aggregate!(1, max).display(true, 200) == "Operation  Col1  Col2  Col3  Col4\n"
+        ~ "max        2     8     7.9   5.6 \n"
+    );
+
+    auto mindf = df.aggregate!(1, min);
+    import std.math: approxEqual;
+    static assert(is(mindf.RowType == AliasSeq!(int, int, double, double)));
+    assert(mindf.data[0][0] == 1 && mindf.data[1][0] == 2);
+    assert(approxEqual(mindf.data[2][0], 3.4, 1e-8) && approxEqual(mindf.data[3][0], 5.6, 1e-8));
+
+    auto doubledf = df.aggregate!(1, max, min);
+    static assert(is(doubledf.RowType == AliasSeq!(int, int, double, double)));
+    assert(doubledf.data[0][0] == 2 && doubledf.data[1][0] == 8);
+    assert(approxEqual(doubledf.data[2][0], 7.9, 1e-8) && approxEqual(doubledf.data[3][0], 5.6, 1e-8));
+    assert(doubledf.data[0][1] == 1 && doubledf.data[1][1] == 2);
+    assert(approxEqual(doubledf.data[2][1], 3.4, 1e-8) && approxEqual(doubledf.data[3][1], 5.6, 1e-8));
+}
+
+// Aggregate Operation on DataFrame Columns
+unittest
+{
+    DataFrame!(int, 2, double, 2) df;
+    Index inx;
+
+    inx[0] = ["Row1", "Row2"];
+    inx[1] = ["Col1", "Col2", "Col3", "Col4"];
+
+    df.setFrameIndex(inx);
+    df = [[1, 2, 3.4, 5.6], [2, 8, 7.9, 5.6]];
+    // df.display();
+
+    import std.algorithm: max, min;
+    import std.math: approxEqual;
+
+    auto maxdf = df.aggregate!(0, max);
+    assert(approxEqual(maxdf.data[0][0], 5.6, 1e-8) && approxEqual(maxdf.data[0][1], 8, 1e-8));
+
+    auto mindf = df.aggregate!(0, min);
+    assert(approxEqual(mindf.data[0][0], 1, 1e-8) && approxEqual(mindf.data[0][1], 2, 1e-8));
+
+    auto doubledf = df.aggregate!(0, min, max);
+    assert(approxEqual(doubledf.data[0][0], 1, 1e-8) && approxEqual(doubledf.data[0][1], 2, 1e-8));
+    assert(approxEqual(doubledf.data[1][0], 5.6, 1e-8) && approxEqual(doubledf.data[1][1], 8, 1e-8));
+}
+
+private auto customFunc(T)(T[] arr)
```

Comment:
move this inside the unittest that uses it and make it static

Reply:
Done in e30f57a :+1: